### PR TITLE
feat(connectors): inbound attachments for Signal + Telegram (#216, PR4/4)

### PR DESCRIPTION
**Pull request by eumemic**
## Summary

PR4 of 4 stacked PRs implementing #216.  Ports Signal and Telegram connectors to surface inbound photos / voice notes / documents / video / audio as SDK ``Attachment`` records on ``emit_inbound``.  The supervisor staging (PR1) + vision pipeline (PR2) take over from there.

**Stacked on:** [#219 (PR3/4)](https://github.com/eumemic/aios/pull/219) — this PR's base is ``wt/recursive-weaving-sundae-pr3``.

### Signal
- ``parse.Attachment`` gains ``host_path`` from signal-cli's envelope ``file`` field — bytes never traverse stdio.
- ``build_content_text`` stops stringifying attachments; harness renders them as ``image_url`` parts or text markers.
- Connector builds SDK ``Attachment`` records via ``_build_sdk_attachments`` with per-item ``as_params`` validation (log+skip on rejection).

### Telegram
- ``parse.Attachment`` (new): carries ``file_id`` (Bot API doesn't pre-download).
- ``_extract_attachments``: photo (largest PhotoSize), voice, document, video, audio.  Stickers and animations dropped.
- Caption becomes ``text``; media-only messages get ``text=""`` rather than being filtered.
- ``filters.TEXT`` → ``filters.ALL``; ``parse_message`` filters channel/bot traffic with full message context.
- ``_download_attachments`` parallelises ``bot.get_file()`` + ``download_to_drive()`` via ``asyncio.gather``; narrows to ``TelegramError`` / ``OSError`` and log+skip on per-attachment failure.

### Out of scope (deferred)

Outbound MEDIA tags are deferred to a future PR.  They require server-side ``_meta.aios.session_id`` plumbing in ``harness/tool_dispatch.py`` so the connector subprocess can map model-visible ``/workspace/...`` paths to host paths.  PR3's ``extract_media_tags`` SDK helper is in place; the wire-up lands separately.

## Stack

1. PR1 ([#217](https://github.com/eumemic/aios/pull/217)) — storage + SDK Attachment + supervisor staging + sandbox bind + GC
2. PR2 ([#218](https://github.com/eumemic/aios/pull/218)) — vision pipeline
3. PR3 ([#219](https://github.com/eumemic/aios/pull/219)) — outbound MEDIA-tag SDK helper
4. **PR4 (this)** — Signal/Telegram inbound port

Merge back-to-front.

## Test plan
- [x] ``uv run mypy src`` — 131 files
- [x] ``uv run ruff check`` + ``ruff format --check`` clean
- [x] ``uv run pytest tests/unit -q`` — 1110 passed
- [x] Signal connector — 73 passed (added 2 tests for host_path capture + new build_content_text behavior)
- [x] Telegram connector — 20 passed (added 3 tests + 3 fixtures for caption/voice/document)
- [x] SDK package — 53 passed (no changes to SDK in this PR)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by eumemic — COMMENTED** [review]
Review of PR4 inbound-attachments port. Posting every finding per Tom's review preferences (score = severity, not suppression). Cannot request-changes on own PR — treat the top-level recommendation as request-changes.

## Counts by severity
- Critical (90-100): 1 — README updates not landed (explicit PR scope item).
- Important (80-89): 2 — outdated docs are now actively misleading; no `_download_attachments` test coverage.
- Minor (60-79): 3 — title says PR4/4 (now PR4/5); silent-drop edge case when all attachments fail; SDK re-validation race.
- Nits (<60): 1 — sticker/animation drop is implicit, not commented.

## Top 3
1. **[score 92] Signal + Telegram READMEs were not updated.** Both still document attachments under "Out of scope for v1" and the Signal one still describes the `[attachment: ...]` text marker the PR just deleted. The PR4 scope (per design call) explicitly called these out.
2. **[score 85] No test coverage for `_download_attachments` / `_build_sdk_attachments`.** parse-layer tests are good, but the connector-layer flow (mock `bot.get_file` + `download_to_drive`, mock signal-cli envelope through to SDK Attachment construction with a real on-disk file so `as_params()` runs end-to-end) is untested. Item 9 of the review brief.
3. **[score 82] Telegram README line 81 is now actively wrong.** "Inbound media (photos, voice, documents, stickers) — dropped silently" is the opposite of what `parse.py` now does. Same for Signal README line 102's `[attachment: …]` marker description.

## Other findings (line comments below)
- **[score 65] PR title "PR4/4" should be "PR4/5"** given the recent stack split (outbound MEDIA tags moved to PR5).
- **[score 65] Silent-drop edge case:** if `_download_attachments` rejects every attachment on a media-only message (`text == ""`), `emit_inbound` fires with empty content + no attachments — the user message vanishes. Not necessarily a bug, but conflicts with "fail hard, no fallbacks." Worth deciding whether to surface a placeholder or skip the emit entirely.
- **[score 60] SDK double-validates attachments:** the connector `_build_sdk_attachments` / `_download_attachments` calls `as_params()`, then `emit_inbound` calls it again. If a host file is removed in between, the drain task dies and kills the connector subprocess. Race window is small; flagging only because the per-attachment skip pattern in connector code is defeated by the SDK's all-or-nothing re-validation.
- **[score 50] Sticker/animation drop is implicit:** `_extract_attachments` returns no entries for sticker/animation messages, so `parse_message` returns None when there's no caption. PR description says this is intentional; consider a one-line comment in `_extract_attachments` noting it.

## Recommendation
**Request changes** (effective). Top item is a documented PR scope miss (READMEs); the testing gap is significant for code that talks to two external services. Everything else is discuss / minor.

File: connectors/signal/src/aios_signal/parse.py [hunk 1]
```diff
@@ -179,9 +181,4 @@ def parse_envelope(
 
 
 def build_content_text(msg: InboundMessage) -> str:
-    parts: list[str] = []
-    if msg.text:
-        parts.append(msg.text)
-    for a in msg.attachments:
-        parts.append(f"[attachment: {a.filename or '(unnamed)'} ({a.content_type})]")
-    return "\n".join(parts)
+    return msg.text
```
Comment (eumemic) [inline connectors/signal/src/aios_signal/parse.py:191]: [score: 92] PR4 scope explicitly included README updates: both connectors document attachments as out-of-scope, and that should be removed (replaced with supported language). Signal README line 102 still describes the `[attachment: <name> (<mime>)]` text-marker behavior that this PR just deleted. Telegram README line 81 still says inbound media is dropped silently. Update both READMEs before merge.

File: connectors/telegram/src/aios_telegram/connector.py [hunk 2]
```diff
@@ -181,15 +195,73 @@ async def _drain_queue(self) -> None:
...
+                )
+                continue
+            out.append(candidate)
+        return out
+
+    async def _download_one(self, att: Attachment) -> Path | None:
+        assert self._application is not None
+        try:
+            file = await self._application.bot.get_file(att.file_id)
+        except TelegramError as err:
```
Comment (eumemic) [inline connectors/telegram/src/aios_telegram/connector.py:247]: [score: 85] No connector-level test for the download flow. The parse-side tests are good, but `_download_attachments` and `_download_one` (the bot.get_file → tempfile → download_to_drive path, plus per-attachment AttachmentError log+skip) are uncovered. Suggest a test_connector.py that mocks `Application.bot.get_file` to return a stub File whose `download_to_drive` writes a small payload to disk, then asserts emit_inbound is called with an SDKAttachment whose host_path is readable. Same gap on the Signal side for `_build_sdk_attachments`.

File: connectors/telegram/src/aios_telegram/connector.py [hunk 3]
```diff
@@ -181,15 +195,73 @@ async def _drain_queue(self) -> None:
...
                 else None
             )
+            sdk_attachments = await self._download_attachments(msg.attachments)
             await self.emit_inbound(
                 account=str(self._bot_id),
                 chat_id=str(msg.chat_id),
                 sender=sender_payload,
                 content=msg.text,
+                attachments=sdk_attachments or None,
                 metadata=metadata,
```
Comment (eumemic) [inline connectors/telegram/src/aios_telegram/connector.py:212]: [score: 65] Edge case: when every attachment fails download/validation on a media-only message (`text == ""`), `sdk_attachments or None` becomes None and `emit_inbound` fires with empty content + no attachments. The user's message effectively vanishes. Per CLAUDE.md "fail hard, no fallbacks" this should probably either skip the emit entirely (drop the inbound, log it loud) or surface a synthetic placeholder. The current behavior is the worst of both worlds — silent partial-success.

File: connectors/telegram/src/aios_telegram/connector.py [hunk 4]
```diff
@@ -181,15 +195,73 @@ async def _drain_queue(self) -> None:
...
+        host_paths = await asyncio.gather(*(self._download_one(a) for a in attachments))
+        out: list[SDKAttachment] = []
+        for att, host_path in zip(attachments, host_paths, strict=True):
+            if host_path is None:
+                continue
+            candidate = SDKAttachment(
+                host_path=str(host_path),
+                filename=att.filename,
+                content_type=att.content_type,
+            )
```
Comment (eumemic) [inline connectors/telegram/src/aios_telegram/connector.py:229]: [score: 60] Double validation: `_download_attachments` calls `candidate.as_params()` to filter rejects, then `emit_inbound` calls `as_params()` again on every attachment in the list. If a host file disappears between the two stat()s the drain loop dies with an unhandled AttachmentError and takes the connector subprocess with it. The race window is tiny but the per-attachment skip in connector code is meaningless given the SDK's all-or-nothing re-validation. Either drop the connector-side validation (let `emit_inbound` raise and handle there) or have the SDK skip-and-log on per-attachment failure. Same shape applies to Signal's `_build_sdk_attachments`.

File: connectors/telegram/src/aios_telegram/parse.py [hunk 5]
```diff
@@ -50,20 +53,71 @@ def _chat_kind(chat_type: str) -> ChatKind:
     return "channel"
 
 
+def _extract_attachments(message: Message) -> tuple[Attachment, ...]:
```
Comment (eumemic) [inline connectors/telegram/src/aios_telegram/parse.py:56]: [score: 50] Sticker/animation drop is silent here — `_extract_attachments` simply doesn't enumerate them, so a sticker-only message returns an empty attachments tuple and parse_message returns None when there's no caption. The PR description says this is intentional; please drop a one-line comment in `_extract_attachments` ("# Stickers and animations not forwarded — out of scope.") so the next reader doesn't think it's an oversight.

**Comment by eumemic** [comment]
[score: 65] PR title says **PR4/4** but the stack was recently split — outbound MEDIA tags moved to a follow-up PR5. Title and body should read **PR4/5** (and the stack list at the top of the body should reflect 5 PRs). Small but worth fixing for the merge log.
